### ISSUE TITLE
[Feature request] prospects of integrating jax's jacfwd, jacrev, and hessian functions in mlx

### ISSUE THREAD

**Issue by LiPingYen** [ml-explore/mlx#164]
### Discussed in https://github.com/ml-explore/mlx/discussions/154

<div type='discussions-op-text'>

<sup>Originally posted by **LiPingYen** December 14, 2023</sup>
In JAX, there are `jacfwd`, `jacrev,` and `hessian` functions for transforming the objective function into functions that compute first-order or second-order derivatives. I'm curious to know if MLX has plans to incorporate these three functions in the future.</div>

**Comment by awni** [comment]
Definitely would like to have these transforms. I will make it a feature request. Contributions for this are very welcome until we have the bandwidth to get to it.

**Comment by LiPingYen** [comment]
Hi @awni ,

I appreciate your responsiveness. Looking forward to leveraging the capabilities of XLM to execute `jacfwd`, `jacrev,` and `hessian` functionalities in the near future.

**Comment by ParamThakkar123** [comment]
Do we want these functions implemented in both python as well C++ codes ??

**Comment by LiPingYen** [comment]
Hi @ParamThakkar123, I think it will be great to implement `jacfwd`, `jacrev`, and `hessian` in python and C++.
Thanks for being the contributor to this enhancement.

**Comment by ParamThakkar123** [comment]
Sure. Going forward with it. Would come up with the enhancement soon

**Comment by LiPingYen** [comment]
Hi @ParamThakkar123, will you sill work on this implementation?

**Comment by ParamThakkar123** [comment]
Yeah, would like to work on this again but would take some time to be back on this

**Comment by LiPingYen** [comment]
I see, thanks @ParamThakkar123 .
